explorer/interpreter/action_stack.cpp

```diff
@@ -192,32 +256,72 @@ auto ActionStack::RunAgain() -> ErrorOr<Success> {
   return Success();
 }
 
-auto ActionStack::UnwindTo(Nonnull<const Statement*> ast_node)
-    -> ErrorOr<Success> {
+auto ActionStack::UnwindTo(Nonnull<const Statement*> ast_node,
+                           bool destroy_scopes) -> ErrorOr<Success> {
   while (true) {
     if (const auto* statement_action =
             llvm::dyn_cast<StatementAction>(todo_.Top().get());
         statement_action != nullptr &&
         &statement_action->statement() == ast_node) {
       break;
     }
-    todo_.Pop();
+    auto item = todo_.Pop();
+    auto& scope = item->scope();
+    if (scope && item->kind() != Action::Kind::CleanUpAction) {
+      std::unique_ptr<Action> cleanup_action =
+          std::make_unique<CleanupAction>(std::move(*scope));
+      destroy_actions_.push_front(std::move(cleanup_action));
+    }
+  }
+  if (destroy_scopes) {
+    for (auto& x : destroy_actions_) {
+      todo_.Push(std::move(x));
+    }
+    destroy_actions_.clear();
   }
   return Success();
 }
 
-auto ActionStack::UnwindPast(Nonnull<const Statement*> ast_node)
-    -> ErrorOr<Success> {
+auto ActionStack::UnwindPast(Nonnull<const Statement*> ast_node,
+                             bool destroy_scopes) -> ErrorOr<Success> {
   CARBON_RETURN_IF_ERROR(UnwindTo(ast_node));
-  todo_.Pop();
-  PopScopes();
+  auto item = todo_.Pop();
+  auto popped_scopes = PopScopes();
+  for (auto& popped_scope : popped_scopes) {
+    destroy_actions_.push_front(std::move(popped_scope));
+  }
+  destroy_actions_.push_front(std::move(item));
+  if (destroy_scopes) {
+    for (auto& x : destroy_actions_) {
+      auto& scope = x->scope();
+      // if(scope && x->kind() != Action::Kind::CleanUpAction){
+      if (scope) {
+        // std::unique_ptr<Action> cleanup_action =
+        // std::make_unique<CleanupAction>(std::move(*scope));
+        // todo_.Push(std::move(cleanup_action));
+        todo_.Push(std::move(x));
+      }
```

Comment:
Should the commented-out code here be removed?

Reply:
Removed!